read-fonts/src/tables/postscript/scale.rs

```diff
@@ -0,0 +1,541 @@
+    pub fn outline(
+        &self,
+        subfont: &ScalerSubfont,
+        glyph_id: GlyphId,
+        coords: &[F2Dot14],
+        pen: &mut impl Pen,
+    ) -> Result<(), Error> {
+        use super::charstring;
+        let charstring_data = self
+            .top_dict
+            .charstrings
+            .as_ref()
+            .ok_or(Error::MissingCharstrings)?
+            .get(glyph_id.to_u16() as usize)?;
+        let subrs = subfont.subrs(self)?;
+        let blend_state = subfont.blend_state(self, coords)?;
+        let mut pen_sink = charstring::PenSink::new(pen);
+        let mut simplifying_adapter = charstring::SimplifyingSink::new(&mut pen_sink);
+        let scale = if subfont.size <= 0.0 {
+            Fixed::ONE
+        } else {
+            // Note: we do an intermediate scale to 26.6 to ensure we
+            // match FreeType
+            Fixed::from_bits((subfont.size * 64.) as i32)
+                / Fixed::from_bits(self.units_per_em as i32)
+        };
+        let mut scaling_adapter =
+            charstring::ScalingSink26Dot6::new(&mut simplifying_adapter, scale);
+        // TODO: hinting will be another sink adapter that slots in here
+        charstring::evaluate(
+            charstring_data,
+            self.global_subrs(),
+            subrs,
+            blend_state,
+            &mut scaling_adapter,
+        )?;
+        simplifying_adapter.finish();
+        Ok(())
+    }
+
+    fn offset_data(&self) -> FontData<'a> {
+        match &self.version {
+            Version::Version1(cff1) => cff1.offset_data(),
+            Version::Version2(cff2) => cff2.offset_data(),
+        }
+    }
+
+    fn global_subrs(&self) -> Index<'a> {
+        match &self.version {
+            Version::Version1(cff1) => cff1.global_subrs().into(),
+            Version::Version2(cff2) => cff2.global_subrs().into(),
+        }
+    }
+
+    fn private_dict_range(&self, subfont_index: u32) -> Result<Range<usize>, Error> {
+        if let Some(font_dicts) = &self.top_dict.font_dicts {
+            // If we have a font dict index, use that
+            let font_dict_data = font_dicts.get(subfont_index as usize)?;
+            let mut range = None;
+            for entry in dict::entries(font_dict_data, None) {
+                if let dict::Entry::PrivateDictRange(r) = entry? {
+                    range = Some(r);
+                    break;
+                }
+            }
+            range
+        } else {
+            // Otherwise, assume the top dict provided a private dict range
+            self.top_dict.private_dict_range.clone()
+        }
+        .ok_or(Error::MissingPrivateDict)
+    }
+}
+
+enum Version<'a> {
+    Version1(Cff<'a>),
+    Version2(Cff2<'a>),
+}
+
+/// Specifies local subroutines and hinting parameters for some subset of
+/// glyphs in a CFF or CFF2 table.
+///
+/// This type is designed to be cacheable to avoid re-evaluating the private
+/// dict every time a charstring is processed.
```

Comment:
again we seem to be presuming our user will cache. Where would one find instructions on implementing a correct, high performance, implementation and ... shouldn't we just give them one?

Reply:
Aha! That's at the very bottom. I think perhaps we could say here that we anticipate most users will use this via skrifa which will provide the caching?

Reply:
I’ll add a note to the Scaler docs

Reply:
Done